src/restruct_poc/beams.hpp

```diff
@@ -0,0 +1,609 @@
+    }
+
+    //--------------------------------------------------------------------------
+    // Populate shape function interpolation/derivative matrices
+    //--------------------------------------------------------------------------
+
+    std::vector<double> weights;
+
+    // Get vector of quadrature points
+    auto qp_xi = input.quadrature.GetQuadraturePoints();
+
+    // Loop through quadrature points
+    for (size_t j = 0; j < qp_xi.size(); ++j) {
+        // Get interpolation weights to go from nodes to this QP
+        LagrangePolynomialInterpWeights(qp_xi[j], node_xi, weights);
+
+        // Copy interp weights to host matrix
+        for (size_t k = 0; k < node_xi.size(); ++k) {
+            shape_interp(k, j) = weights[k];
+        }
+
+        // Get derivative weights to go from nodes to this QP
+        LagrangePolynomialDerivWeights(qp_xi[j], node_xi, weights);
+
+        // Copy deriv weights to host matrix
+        for (size_t k = 0; k < node_xi.size(); ++k) {
+            shape_deriv(k, j) = weights[k];
+        }
+    }
+
+    //--------------------------------------------------------------------------
+    // Interpolate section mass and stiffness matrices to quadrature points
+    // TODO: remove assumption that s runs from 0 to 1.
+    //--------------------------------------------------------------------------
+
+    // Get section positions [-1,1]
+    std::vector<double> section_xi(input.sections.size());
+
+    // Loop through sections and convert 's' [0,1] position to xi
+    for (size_t i = 0; i < input.sections.size(); ++i) {
+        section_xi[i] = 2 * input.sections[i].s - 1;
+    }
+
+    // Fill view with zeros
+    Kokkos::deep_copy(qp_Mstar, 0.);
+    Kokkos::deep_copy(qp_Cstar, 0.);
+
+    // Loop through quadrature points and calculate section weights for interp
+    for (size_t i = 0; i < qp_xi.size(); ++i) {
+        // Calculate weights
+        LagrangePolynomialInterpWeights(qp_xi[i], section_xi, weights);
+
+        // Loop through sections
+        for (size_t j = 0; j < section_xi.size(); ++j) {
+            for (size_t m = 0; m < 6; ++m) {
+                for (size_t n = 0; n < 6; ++n) {
+                    qp_Mstar(i, m, n) += input.sections[j].M_star[m][n] * weights[j];
+                    qp_Cstar(i, m, n) += input.sections[j].C_star[m][n] * weights[j];
+                }
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------
+// Beams data structure
+//------------------------------------------------------------------------------
+
+struct Beams {
+    size_t num_beams_;  // Number of beams
+    size_t num_nodes_;  // Number of nodes
+    size_t num_qps_;    // Number of quadrature points
+
+    Kokkos::View<BeamElemIndices*> elem_indices;  // View of element node and qp indices into views
+    Kokkos::View<int*> node_state_indices;        // State row index for each node
+
+    oturb::View_3 gravity;
```

Comment:
I get that gravity is the same for all nodes, but isn't this just an applied force at each node?  Does gravity need its own vector, or can it just be included there at assembly time?

Reply:
Gravity is used in the calculation of `Fg` at the quadrature points, which is based on the rotated mass matrix. Since this is part of the quadrature point calculation, and is integrated back to the nodes, I thought it made sense to have it as a view. It could also be passed into the `UpdateStates` routine, but then it'd have to be transferred to the GPU before it could be used.

Reply:
Yea, I was thinking it would be a force that would be handled by our integration routine at the same level as a point load or applied aerodynamic load.  I'm not sure there's much reason to handle gravity any differently and it will give us another simple example when working out that interface.